src/components/text-editor/prosemirror-adapter/plugins/trigger-plugin.ts

```diff
@@ -0,0 +1,352 @@
+// import { ListItem } from 'src/components/list/list-item.types';
+
+import { Node, NodeType, Schema } from 'prosemirror-model';
+import { Plugin, PluginKey, EditorState } from 'prosemirror-state';
+import { ReplaceStep, ReplaceAroundStep } from 'prosemirror-transform';
+import { EditorView } from 'prosemirror-view';
+import { CustomElement } from 'src/interface';
+
+export const triggerPluginKey = new PluginKey('triggerPlugin');
+
+export type TriggerCharacter =
+    | '@'
+    | '#'
+    | '$'
+    | '!'
+    | '?'
+    | '&'
+    | '*'
+    | '%'
+    | '+'
+    | '-'
+    | '='
+    | '/'
+    | '\\'
+    | '^'
+    | '~'
+    | '`'
+    | ':'
+    | ';'
+    | '|'
+    | '.'
+    | ','
+    | '<'
+    | '>'
+    | '['
+    | ']'
+    | '{'
+    | '}'
+    | '('
+    | ')'
+    | "'";
+
+export type CustomElementInstance = Omit<CustomElement, 'attributes'> & {
+    attributes: Record<string, any>;
+};
+
+export type TextEditorNode = {
+    node: CustomElementInstance | string;
+    children?: Array<TextEditorNode | string>;
+};
+
+export interface TextEditor {
+    insert: (input: TextEditorNode | string) => void;
+}
+
+export interface TriggerEventDetail {
+    /**
+     * The trigger that triggered this event
+     */
+    trigger: TriggerCharacter;
+
+    /**
+     * The text editor
+     */
+    textEditor: TextEditor;
+
+    /**
+     * Current value of the trigger
+     */
+    value: string;
+}
+
+export const inserterFactory = (view: EditorView): TextEditor => {
```

Comment:
Have we investigated if there are any existing trigger functionality we can import instead of re-inventing the wheel?